apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -0,0 +1,487 @@
+            }
+        }
+
+        ///<inheritdoc/>
+        public int CompareTo(PgNumeric other)
+        {
+            // Check for NaN.
+            // In order to allow numeric values to be sorted and used in tree-based indexes,
+            // PostgreSQL treats NaN values as equal, and greater than all non-NaN values.
+            if (_notANumber)
+            {
+                return other._notANumber ? 0 : 1;
+            }
+
+            // if scale is same, we can compare the integer value.
+            if (_scale == other._scale)
+            {
+                return _scaledValue.CompareTo(other._scaledValue);
+            }
+
+            // return comparison of normalized value.
+            return GetNormalizedValue().CompareTo(other.GetNormalizedValue());
+        }
+
+        /// <summary>
+        /// Implementation of <see cref="IComparable.CompareTo"/> to compare two <see cref="PgNumeric"/> values.
+        /// </summary>
+        /// <remarks>
+        /// This uses explicit interface implementation to avoid it being called accidentally. The generic implementation should usually be preferred.
+        /// </remarks>
+        /// <exception cref="ArgumentException"><paramref name="obj"/> is non-null but does not refer to an instance of <see cref="PgNumeric"/>.</exception>
+        /// <param name="obj">The object to compare this value with.</param>
+        /// <returns>The result of comparing this value with <paramref name="obj"/>. <paramref name="obj"/> is null, this method returns a value greater than 0.
+        /// </returns>
+        int IComparable.CompareTo(object obj)
+        {
+            if (obj == null)
+            {
+                return 1;
+            }
+            return obj is PgNumeric numeric
+                ? CompareTo(numeric)
+                : throw new ArgumentException($"Object must be of type {nameof(PgNumeric)}", nameof(obj));
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(PgNumeric other) => CompareTo(other) == 0;
+
+        /// <summary>
+        /// Gets the normalized value of the PgNumeric instance.
+        /// </summary>
+        /// <returns>The normalized value as big integer.</returns>
+        public BigInteger GetNormalizedValue()
+        {
+            int differenceToMaxScale = MaxScale - _scale;
+            if (!s_powersOf10.TryGetValue(differenceToMaxScale, out BigInteger powerOf10))
+            {
+                powerOf10 = BigInteger.Pow(10, differenceToMaxScale);
+                s_powersOf10.Add(differenceToMaxScale, powerOf10);
+            }
+
+            return _scaledValue * powerOf10;
+        }
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj) => obj is PgNumeric other && Equals(other);
+
+        /// <inheritdoc/>
+        public override int GetHashCode() => GetNormalizedValue().GetHashCode() + _notANumber.GetHashCode();
```

Comment:
I'd suggest using a fixed "large and negative" value for the hash code of `_notANumber`, otherwise it'll end up being 1 which is more likely to cause hash collisions.

It's a bit of a shame that we need to call GetNormalizedValue here, when hashing should be cheap.